### PR TITLE
ISPN-7889 BaseDistributionInterceptor.remoteGet may cause concurrency issues

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-7889
https://issues.jboss.org/browse/ISPN-8889

* Rename remoteGet to remoteGetSingleKey and use it only for single-key
  commands.
* Rename remoteGetAll to remoteGetMany and use it for all multi-key
  commands.
* Throw OutdatedTopologyException.RETRY_SAME_TOPOLOGY after receiving
  only UnsureResponses and RETRY_NEXT_TOPOLOGY if there was at least one
  CacheNotFoundResponse (and no valid response).
* Make remote scattered reads wait for the right topology
* Improve support for PartitionHandling.ALLOW_READS in scattered caches.
* Make partition availability checks more readable.
* Add asynchronous methods in ControlledRpcManager
* Introduce OutdatedTopologyException.RETRY_SAME_TOPOLOGY

I've separated the OutdatedTopologyException changes into a commit, they touch quite a lot of files. I can do the same for the ControlledRpcManager, I changed some method names there and it affected quite a few tests.

I need your review @rvansa, especially for the scattered mode changes. I tried to add/clarify comments as much as possible, but I might have missed some stuff ;)

### CODE REVIEW

File: core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java [hunk 1]
```diff
@@ -248,33 +250,76 @@ protected void doCheck(Object key, boolean isWrite, long flagBitSet) {
...
+
+   protected boolean isBulkOperationAllowed() {
+      if (availabilityMode == AvailabilityMode.AVAILABLE)
+         return true;
+
+      switch (partitionHandling) {
+         case ALLOW_READ_WRITES:
+            return true;
+         case ALLOW_READS:
+            // Never allow bulk reads/writes
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java]: Why not allowing bulk reads? (please add comment)

File: core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java [hunk 2]
```diff
@@ -96,6 +96,7 @@ public void notifyTransactionDataReceived(int topologyId) {
          if (transactionDataTopologyId >= expectedTopologyId) {
             return CompletableFutures.completedNull();
          } else {
+            // TODO Dan: Use thenComposeAsync to continue each command in a separate thread?
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java]: No, the thread in which the execution should continue should be decided by the caller of `transactionDataFuture`.

File: core/src/test/java/org/infinispan/distribution/DistAsyncFuncTest.java [hunk 3]
```diff
@@ -65,7 +66,8 @@ protected void createCacheManagers() throws Throwable {
             protected <T> CompletionStage<T> performRequest(Collection<Address> targets, ReplicableCommand command,
                                                             ResponseCollector<T> collector,
                                                             Function<ResponseCollector<T>,
-                                                               CompletionStage<T>> invoker) {
+                                                                       CompletionStage<T>> invoker,
```
Comment (rvansa) [inline core/src/test/java/org/infinispan/distribution/DistAsyncFuncTest.java]: indentation?

File: core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java [hunk 4]
```diff
@@ -79,11 +79,11 @@ public Object visitGetKeysInGroupCommand(InvocationContext ctx, GetKeysInGroupCo
          return invokeNextThenAccept(ctx, command, (rCtx, rCommand, rv) -> {
             GetKeysInGroupCommand cmd = (GetKeysInGroupCommand) rCommand;
             final int commandTopologyId = cmd.getTopologyId();
+            // TODO Dan: We used to throw an exception if the node was still a write owner
+            //  The new check should be better, but maybe we should retry if it's not a primary owner any more?
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java]: Ok, I'll remove the comment

File: core/src/main/java/org/infinispan/statetransfer/OutdatedTopologyException.java [hunk 5]
```diff
@@ -16,31 +17,36 @@
...
+   /**
+    * A cached instance, used for read commands that need to be retried in the same topology.
+    *
+    * <p>This happens because we read from backup owners when the primary owners no longer have the entry,
+    * so we only retry when all of the owners reply with an UnsureResponse.
+    * Topologies T and T+1 always have at least one read owner in common, so receiving UnsureResponse from all the
+    * owners means either one owner had topology T+2 and by now we have at least T+1, or one owner had topology T-1
+    * and another had T+1, and by now all should have at least T.</p>
+    */
+   public static final OutdatedTopologyException RETRY_SAME_TOPOLOGY =
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/OutdatedTopologyException.java]: I just ran the inspection and it didn't complain, so I'm deleting the annotation from the other

File: core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java [hunk 6]
```diff
@@ -192,32 +202,34 @@ private Object handleExceptionOnCommandReturn(InvocationContext rCtx, VisitableC
...
+
+         // We can get OTE for dist reads even if current topology information is sufficient:
+         // 1. A has topology in phase READ_ALL_WRITE_ALL, sends message to both old owner B and new C
+         // 2. C has old topology with READ_OLD_WRITE_ALL, so it responds with UnsureResponse
+         // 3. C updates topology to READ_ALL_WRITE_ALL, B updates to READ_NEW_WRITE_ALL
+         // 4. B receives the read, but it already can't read: responds with UnsureResponse
+         // 5. A receives two unsure responses and throws OTE
+         // However, now we are sure that we can immediately retry the request, because C must have updated its topology
+         OutdatedTopologyException ote = (OutdatedTopologyException) ce;
+         if (ote.requestedTopologyId >= 0) {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java]: The trick (was) is that I wanted to allow exceptions with a custom topology id as well, but I've removed those usages so I'll try to simplify the check here

File: core/src/main/java/org/infinispan/interceptors/distribution/RemoteGetSingleKeyCollector.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+   public SuccessfulResponse addResponse(Address sender, Response response) {
+      if (response.isSuccessful()) {
+         return (SuccessfulResponse) response;
+      }
+      if (response instanceof ExceptionResponse) {
+         throw ResponseCollectors.wrapRemoteException(sender, ((ExceptionResponse) response).getException());
+      }
+
+      if (response instanceof UnsureResponse) {
+         hasUnsureResponse = true;
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/distribution/RemoteGetSingleKeyCollector.java]: Yep, removed

File: core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java [hunk 8]
```diff
@@ -1061,4 +919,118 @@ protected Object invokeRemoveExpiredCommand(InvocationContext ctx, RemoveExpired
...
+
+         List<Object> senderKeys = requestedKeys.get(sender);
+         InternalCacheValue[] values = (InternalCacheValue[]) responseValue;
+         for (int i = 0; i < senderKeys.size(); ++i) {
+            Object key = senderKeys.get(i);
+            InternalCacheValue value = values[i];
+            CacheEntry entry = value == null ? NullCacheEntry.getInstance() : value.toInternalCacheEntry(key);
+            wrapRemoteEntry(ctx, key, entry, false);
+         }
+         // TODO Dan: transform the remote entries before wrapping them in the context
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java]: What I was trying to say is that `handleRemotelyRetrievedKeys` could receive a list of remote values and insert them into the context itself

File: core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java [hunk 9]
```diff
@@ -313,16 +312,26 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx,
          InvocationContext ctx, C command, WriteManyCommandHelper<C, Container, Item> helper, ConsistentHash ch,
          MergingCompletableFuture<Object> allFuture, MutableInt offset, IntSet segments) throws Exception {
       Container myItems = helper.newContainer();
-      List<CompletableFuture<?>> retrievals = null;
+      List<Object> remoteKeys = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java]: Ok

File: core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java [hunk 10]
```diff
@@ -335,12 +344,11 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx,
...
       InvocationFinallyAction handler =
             createLocalInvocationHandler(ch, allFuture, segments, helper, MergingCompletableFuture.moveListItemsToFuture(myOffset));
-      if (retrievals == null) {
+      if (remoteKeys.isEmpty()) {
          invokeNextAndFinally(ctx, localCommand, handler);
       } else {
          // We must wait until all retrievals finish before proceeding with the local command
-         CompletableFuture[] ra = retrievals.toArray(new CompletableFuture[retrievals.size()]);
-         Object result = asyncInvokeNext(ctx, command, CompletableFuture.allOf(ra));
+         Object result = asyncInvokeNext(ctx, command, retrievals);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java]: @rvansa Ouch, I read lines 348 and 352 invoking the next interceptor and I almost had a heart attack thinking that we're going to have multiple threads accessing the context :)

`CallInterceptor` should be pretty fast, wouldn't it simplify things if we only invoked it after we got all the entries in the context?

File: core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java [hunk 11]
```diff
@@ -384,16 +392,25 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx,
 
    private <C extends WriteCommand, Item> Object handleRemoteReadWriteManyCommand(
          InvocationContext ctx, C command, WriteManyCommandHelper<C, ?, Item> helper) throws Exception {
-      List<CompletableFuture<?>> retrievals = null;
+      List<Object> remoteKeys = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java]: Ok

File: core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java [hunk 12]
```diff
@@ -192,32 +202,34 @@ private Object handleExceptionOnCommandReturn(InvocationContext rCtx, VisitableC
...
+      int requestedTopologyId;
       if (ce instanceof SuspectException) {
          if (trace)
             getLog().tracef("Retrying command because of suspected node, current topology is %d: %s",
                   currentTopologyId, rCommand);
-         // It is possible that current topology is actual but the view still contains a node that's about to leave;
-         // a broadcast to all nodes then can end with suspect exception, but we won't get any new topology.
-         // An example of this situation is when a node sends leave - topology can be installed before the new view.
-         // To prevent suspect exceptions use SYNCHRONOUS_IGNORE_LEAVERS response mode.
+         requestedTopologyId = cmd.getTopologyId() + 1;
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/impl/BaseStateTransferInterceptor.java]: This branch should never be used, read commands should never throw a `SuspectException`. 
I'll rename the method to make it clearer that it's only used for read commands and simplify the branch.

File: core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java (hunk 1 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/partitionhandling/impl/PartitionHandlingManagerImpl.java]: Good question... I think we can allow bulk reads for replicated caches :)

File: core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java (hunk 2 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java]: Good point, so we just need to modify `DefaultTopologyRunnable` to use `thenApplyAsync`... or remove `DefaultTopologyRunnable` altogether :)

File: core/src/test/java/org/infinispan/distribution/DistAsyncFuncTest.java (hunk 3 above)
Comment (danberindei) [reply core/src/test/java/org/infinispan/distribution/DistAsyncFuncTest.java]: That's IntelliJ aligns it for me if it's on a separate line, but it fits on the `Function<` line so I moved it

**Comment by danberindei** [comment]
Updated @rvansa @wburns

File: core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java]: I made the list allocation lazy everywhere.

File: core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java (hunk 2 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java]: `BaseStateTransferInterceptor` also uses `transactionDataReceived()` directly, so I created https://issues.jboss.org/browse/ISPN-9852

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !

**Pull request merged**